### ISSUE TITLE
S2583 and S2589 False positive when a variable is initialized inside local function

### ISSUE THREAD

**Issue by volgunin** [SonarSource/sonar-dotnet#3161]
### Description

When a variable is initialized inside local function, S2583 is raised.

### Repro steps
```
public void Test()
{
        string buffer = null;
        Local();
        if (buffer != null)
        {
        }

        void Local()
        {
             buffer = new string('a', 1);
        }
}
```
### Expected behavior

No warning.

### Actual behavior

warning S2583: Change this condition so that it does not always evaluate to 'false'; some subsequent code is never executed.

### Related information

SonarAnalyzer.CSharp.8.4.0.15306
VisualStudio 2019 16.4.4

**Comment by Corniel** [comment]
I think your example is to simple, as now you're evaluation always validate to be true. ;)

That being said, the assigning via the private helper method is indeed not detected properly, the construct in itself should be strongly discouraged, as it extremely hard to read/maintain.

``` C#
public class S2583
{
    public void Test()
    {
        string buffer = null;
        PrivateHelper();
        if (buffer != null)
        {
            throw new Exception();
        }

        void PrivateHelper()
        {
            var rnd = new Random();
            if (rnd.Next() == 0)
            {
                buffer = new string('a', 1);
            }
        }
    }
}
```

**Comment by andrei-epure-sonarsource** [comment]
thanks @volgunin and @Corniel for the discussion.

While I agree this is a FP, it's indeed quite an anti-pattern and it can be easily solved by returning the value from the local function.

I am tempted to mark it as won't fix. @pavel-mikula-sonarsource what's your view on this?

**Comment by pavel-mikula-sonarsource** [comment]
I can imagine some `ResetState()` scenario heavily (3+) depending on captured scope where this can be actually useful. And it should get fixed as side effect of bulletproof CFG/SE.

I vote for keeping this issue and seeing if CFG/SE future improvements fixes this as well.
